### PR TITLE
Record SLM history into an index

### PR DESCRIPTION
**Pull request by AthenaEryma**
It is useful to have a record of the actions that Snapshot Lifecycle
Management takes, especially for the purposes of alerting when a
snapshot fails or has not been taken successfully for a certain amount of
time.

This adds the infrastructure to record SLM actions into an index that
can be queried at leisure, along with a lifecycle policy so that this
history does not grow without bound.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotHistoryStore.java
Comment (gwbrown) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotHistoryStore.java]: This boolean is not currently consumed, I added it intending to put results in a bounded queue if the templates haven't been installed yet, but that might not be worth the extra complexity.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotHistoryStore.java
Comment (gwbrown) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotHistoryStore.java]: This is written to use monthly indices at the moment because it's likely that this index will have a very low volume of writes - I expect ~1/day to be common, and even the relatively high-volume usage of taking a snapshot every 30m would only produce a maximum of ~1500/month. Even 10 times that is still a small number of documents for one index.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotCreationHistoryItem.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotCreationHistoryItem.java]: Javadocs please :)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotCreationHistoryItem.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotCreationHistoryItem.java]: I assume this is for later when we have a `DELETE` operation also with retention?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotCreationHistoryItem.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotCreationHistoryItem.java]: It seems strange to tie this to `CreateSnapshotRequest`, I think it'd be better to just pass the repository and snapshot strings in directly. I think it'd be nicer for testing too (since you don't have to make a fake `CreateSnapshotRequest` if you want to store something

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotCreationHistoryItem.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotCreationHistoryItem.java]: Maybe since we have the policy in `SnapshotLifecycleTask`, we should just pass in the time, `policy`, and attempted snapshot name here and derive the other parts from it?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotHistoryItem.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotHistoryItem.java]: Rather than do this, why not have the subclass just override `writeTo` and call `super.writeTo(out)`? Do you just want to ensure that any subclass *has* to override it?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotHistoryStore.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotHistoryStore.java]: Javadocs please :)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistry.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistry.java]: Javadocs :)

File: x-pack/plugin/core/src/main/resources/slm-history.json
Comment (dakrone) [inline x-pack/plugin/core/src/main/resources/slm-history.json]: I think setting this to `strict` would make it easier to catch bugs in our tests

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java
Comment (dakrone) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java]: I don't think we need to mock, we have `TestThreadPool` that can be used instead (it just needs to be shut down at the end of the test so it doesn't leak threads

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java
Comment (dakrone) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java]: I also think we can do this without mocking client, using `NoOpClient` instead, maybe the same way it is in `VerifyingClient` in `SnapshotLifecycleTaskTests`?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java
Comment (dakrone) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java]: We can `ClusterServiceUtils` which can create a real ClusterService that's used for testing so we don't have to mock this

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleTaskTests.java
Comment (dakrone) [inline x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleTaskTests.java]: I think rather than mock the snapshot history store, we could subclass it and verify it with a real method without actually performing the indexing operation (I think mocks tend to make our tests more fragile)

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java
Comment (dakrone) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java]: It looks like most of these are copied from the watcher template registry tests, I'm not sure if it makes sense to have exhaustive tests for each (probably a different discussion).

I also see why there's so much mocking (copied over from the other test implementation). If we keep these mocks that's okay (though I greatly prefer we move away from them), but I understand why they're here. It's up to you whether to leave them or change them

**Review by dakrone — CHANGES_REQUESTED** [review]
Thanks Gordon! I took a look and left some comments on this.

I do have one overall comment though, I think subclassing `SnapshotHistoryItem` is getting a little too over complicated for what it is. I think instead we should collapse `SnapshotCreationHistoryItem` back into `SnapshotHistoryItem` and just have a `"type"` field that is `creation` or (eventually) `deletion`. Not subclassing also means we don't have to try and keep a mapping up to date with a base class plus whatever fields a subclass would happen to add. This should simplify it a bit and make it easier to read also.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java
Comment (gwbrown) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java]: I'm not opposed to changing the mocks - I did a little other code cleanup when bringing these tests over too so I don't see why I shouldn't do that as well.

My position is that if we're going to use a superclass to share code, then the concrete subclasses should all test the behaviors they rely on (which is why I didn't bring over *all* the tests - the Watcher version depends on some behavior that this one doesn't need).  That way, if e.g. Watcher is radically refactored or removed, etc. and the Watcher template registry is removed then these tests will still verify the behavior that SLM relies on.

I would like to factor out some of the shared logic so that it's less copy/paste-y, but I didn't want to start making that kind of change to the Watcher tests in this feature branch. Ideally I would have refactored the tests as well in the PR that factored out the shared superclass, but I didn't want to delay this PR for another round of changes to Watcher code.

File: x-pack/plugin/core/src/main/resources/slm-history.json
Comment (gwbrown) [reply x-pack/plugin/core/src/main/resources/slm-history.json]: It's true that it would make it easier to catch bugs in our tests, but the reason I don't want to set `strict` in this mapping is that:
1) Looking at `CreateSnapshotRequest`, there are some additional options that don't appear to be commonly used and I don't totally understand. For example, the request object [parses out a `settings` field](https://github.com/elastic/elasticsearch/blob/5b97fffb016f9cd5ec88c75711c2f2ad34e6a773/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/create/CreateSnapshotRequest.java#L400-L404) which doesn't appear to be currently used, and [`ignore_unavailable` is actually part of an `IndicesOptions` class that has a number of options](https://github.com/elastic/elasticsearch/blob/2d6c46255cd3f642ddbf18fa55e2fad194fc01b6/server/src/main/java/org/elasticsearch/action/support/IndicesOptions.java#L324-L331) available, which may or may not be in use for snapshots. I don't want to run into issues where these change, or an option gets added, and recording history fails just because we didn't add the field to the mapping (especially since those fields will probably never get queried and so don't need to be indexed, but may be useful to store in the _source).
2) Once `_meta` for snapshots is merged, users will be able to add that to the `config` section of their policy, which can contain arbitrary keys. We might be able to filter it out for this history record, but the simplicity of just storing the entire configuration is nice.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotCreationHistoryItem.java
Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotCreationHistoryItem.java]: This has been switched to use the policy instead of the request as suggested below.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotCreationHistoryItem.java
Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotCreationHistoryItem.java]: Good idea! I've switched these two methods to use the policy.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotHistoryItem.java
Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotHistoryItem.java]: This method is no longer relevant as I've gotten rid of the inheritance as suggested at the top level.

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleTaskTests.java
Comment (gwbrown) [reply x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleTaskTests.java]: Done.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java
Comment (gwbrown) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java]: I've changed these tests to remove all the mocks, instead using ClusterServiceUtils, TestThreadPool, VerifyingClient, etc. as recommended.

**Review by jakelandis — CHANGES_REQUESTED** [review]
Overall looks good !

Wanted to double check on a couple of the test classes if they could be implemented with mockito instead. Ideally we should avoid subclassing real objects for verification purposes if mocks can offer the same level of verification.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotHistoryItem.java
Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotHistoryItem.java]: The only information the snapshot configuration will contain is:
1) the list of included indices,
2) Some boolean flags (`include_global_state`, `partial`, `ignore_unavailable`, etc)
3) Once https://github.com/elastic/elasticsearch/pull/41281 is merged, arbitrary metadata

So the only way this would contain sensitive information is if the user attaches it in the metadata, and there's not much we can do about that.

File: x-pack/plugin/core/src/main/resources/slm-history.json
Comment (gwbrown) [reply x-pack/plugin/core/src/main/resources/slm-history.json]: Good point, I can remove this - I think this is just a holdover from copying the Watcher history template (and Watcher does use this field, but SLM does not).

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java
Comment (gwbrown) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java]: I actually used mocks in the first version of this PR, but in his review Lee recommended pretty strongly that we move away from mocks (https://github.com/elastic/elasticsearch/pull/41707#discussion_r280606852 and https://github.com/elastic/elasticsearch/pull/41707#discussion_r280606426).

I removed the mocks in favor of the subclassing approach in https://github.com/elastic/elasticsearch/pull/41707/commits/483cd516781afbd0515dde84073c1d112a07c1fd and https://github.com/elastic/elasticsearch/pull/41707/commits/ea8b159f9c3d66ed2e0507c905bf426f9624c235 - I don't personally have strong feelings about this, and can see advantages to both approaches.

This seems to be something that warrants more discussion in a higher-bandwidth forum, so would you be okay with this going in with the subclassing for now, pending changes at that discussion? It ought to be pretty easy to convert back to mocks if that's the path we decide we want to take.

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleTaskTests.java
Comment (gwbrown) [reply x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleTaskTests.java]: See comment at https://github.com/elastic/elasticsearch/pull/41707#discussion_r282259103

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java
Comment (jakelandis) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/snapshotlifecycle/history/SnapshotLifecycleTemplateRegistryTests.java]: Sounds good. Let's not block this PR on that discussion.

**Review by jakelandis — APPROVED** [review]
LGTM

**Pull request merged**
